pkg/provisioner/crdprovisioner.go

```diff
@@ -643,9 +642,35 @@ func (c *CrdProvisioner) demoteVolume(ctx context.Context, azVAClient v1beta1.Az
 	updateFunc := func(obj interface{}) error {
 		updateInstance := obj.(*diskv1beta1.AzVolumeAttachment)
 		updateInstance.Spec.RequestedRole = diskv1beta1.ReplicaRole
+
+		if updateInstance.Labels == nil {
+			updateInstance.Labels = map[string]string{}
+		}
+
+		updateInstance.Labels[consts.RoleLabel] = string(diskv1beta1.ReplicaRole)
+		updateInstance.Labels[consts.RoleChangeLabel] = consts.Demoted
 		return nil
 	}
-	return azureutils.UpdateCRIWithRetry(ctx, c.conditionWatcher.informerFactory, nil, c.azDiskClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry)
+	return azureutils.UpdateCRIWithRetry(ctx, c.conditionWatcher.informerFactory, nil, c.azDiskClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateAll)
+}
+
+func (c *CrdProvisioner) DetachVolume(ctx context.Context, volumeID string,
```

Comment:
I find it very confusing to have both an `UnpublishVolume` and `DetachVolume` method. Which do I use when?

I recommend adding an `UnpublishMode` enum parameter to `UnpublishVolume` to tailor its behavior to the scenario. We can start with values `DemoteOrDetach` for the normal case and `Detach` for the new case. Later we can add `ForceDetach` when we understand the scenario in which that could be useful.

I also recommend refactoring the `maxShares` check into a separate internal method `func (c *CrdProvisioner) shouldDemote(volumeName string, mode UnpublishMode) bool` that only gets the `AzVolume` if `mode == DemoteOrDetach`.